Defect: Prepocessor fails to expand x-macro macro parameter

cppcheck fails to preprocess the code below:

```
> cppcheck --version
Cppcheck 2.3
> cppcheck -E main.c
main.c:12:0: error: failed to expand 'COLOR_SET', Wrong number of parameters for macro 'ALL_COLORS'. [preprocessorErrorDirective]
#define COLOR_SET ALL_COLORS(WARM_COLORS)
^

```

main.c:

```
#define ALL_COLORS(warm_colors) \
	X(Blue) \
	X(Green) \
	X(Purple) \
	warm_colors

#define WARM_COLORS \
	X(Red) \
	X(Yellow) \
	X(Orange)

#define COLOR_SET ALL_COLORS(WARM_COLORS)

#define X(color) #color,

COLOR_SET
```

gcc handles it fine:

```
> gcc -E main.c
# 1 "main.c"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "main.c"
# 16 "main.c"
"Blue", "Green", "Purple", "Red", "Yellow", "Orange",
```

--- a/test.cpp
+++ b/test.cpp
@@ -747,6 +747,41 @@ static void define15() // #231
                   "x", preprocess(code));
 }
 
+static void define16() // #201
+{
+    const char code[] = "#define ALL_COLORS(warm_colors) \\\n"
+                        "  X(Blue) \\\n"
+                        "  X(Green) \\\n"
+                        "  X(Purple) \\\n"
+                        "  warm_colors\n"
+                        "\n"
+                        "#define WARM_COLORS \\\n"
+                        "  X(Red) \\\n"
+                        "  X(Yellow) \\\n"
+                        "  X(Orange)\n"
+                        "\n"
+                        "#define COLOR_SET ALL_COLORS(WARM_COLORS)\n"
+                        "\n"
+                        "#define X(color) #color,\n"
+                        "\n"
+                        "COLOR_SET\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\"Blue\" , \"Green\" , \"Purple\" , \"Red\" , \"Yellow\" , \"Orange\" ,", preprocess(code));
+}
 
 
 
@@ -3725,6 +3760,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define13);
     TEST_CASE(define14); // #296
     TEST_CASE(define15); // #231
+    TEST_CASE(define16); // #201
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -811,6 +811,18 @@ static void define17() // #185
                   "static_assert ( true , \" \" ) ;", preprocess(code));
 }
 
+static void define18() // #130
+{
+    const char code[] = "#define MAC2STR(x) x[0],x[1],x[2],x[3],x[4],x[5]\n"
+                        "#define FT_DEBUG(fmt, args...) if(pGlobalCtx && pGlobalCtx->debug_level>=2) printf(\"FT-dbg: \"fmt, ##args)\n"
+                        "\n"
+                        "FT_DEBUG(\"  %02x:%02x:%02x:%02x:%02x:%02x\\n\", MAC2STR(pCtx->wlan_intf_addr[i]));\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "if ( pGlobalCtx && pGlobalCtx -> debug_level >= 2 ) printf ( \"FT-dbg: \" \"  %02x:%02x:%02x:%02x:%02x:%02x\\n\" , pCtx -> wlan_intf_addr [ i ] [ 0 ] , pCtx -> wlan_intf_addr [ i ] [ 1 ] , pCtx -> wlan_intf_addr [ i ] [ 2 ] , pCtx -> wlan_intf_addr [ i ] [ 3 ] , pCtx -> wlan_intf_addr [ i ] [ 4 ] , pCtx -> wlan_intf_addr [ i ] [ 5 ] ) ;", preprocess(code));
+}
+
 
 static void define_invalid_1()
 {
@@ -3789,6 +3801,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define15); // #231
     TEST_CASE(define16); // #201
     TEST_CASE(define17); // #185
+    TEST_CASE(define18); // #130
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -857,6 +857,44 @@ static void define21() // #66
                   "foo ( 1 , ( ( a ) ) + 1 ) ;", preprocess(code));
 }
 
+static void define22() // #40
+{
+    const char code[] = "#define COUNTER_NAME(NAME, ...) NAME##Count\n"
+                        "#define COMMA ,\n"
+                        "\n"
+                        "#define DECLARE_COUNTERS(LIST) unsigned long LIST(COUNTER_NAME, COMMA);\n"
+                        "\n"
+                        "#define ACTUAL_LIST(FUNCTION, SEPARATOR) \\\n"
+                        "FUNCTION(event1, int, foo) SEPARATOR \\\n"
+                        "FUNCTION(event2, char, bar)\n"
+                        "\n"
+                        "DECLARE_COUNTERS(ACTUAL_LIST)\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "unsigned long event1Count , event2Count ;", preprocess(code));
+}
+
+static void define23() // #40
+{
+    const char code[] = "#define COMMA ,\n"
+                        "#define MULTI(SEPARATOR) A SEPARATOR B\n"
+                        "\n"
+                        "#define VARS MULTI(COMMA)\n"
+                        "unsigned VARS;\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "unsigned A , B ;", preprocess(code));
+}
+
 
 static void define_invalid_1()
 {
@@ -3839,6 +3877,8 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define19); // #124
     TEST_CASE(define20); // #113
     TEST_CASE(define21); // #66
+    TEST_CASE(define22); // #40
+    TEST_CASE(define23); // #40
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -835,6 +835,18 @@ static void define19() // #124
                   "const Testsuffix ;", preprocess(code));
 }
 
+static void define20() // #113
+{
+    const char code[] = "#define TARGS4 T1,T2,T3,T4\n"
+                        "#define FOOIMPL(T__CLASS, TARGS) void foo(const T__CLASS<TARGS>& x) { }\n"
+                        "#define FOOIMPL_4(T__CLASS)      FOOIMPL(T__CLASS, TARGS4)\n"
+                        "FOOIMPL_4(y)\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "void foo ( const y < T1 , T2 , T3 , T4 > & x ) { }", preprocess(code));
+}
+
 
 static void define_invalid_1()
 {
@@ -3815,6 +3827,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define17); // #185
     TEST_CASE(define18); // #130
     TEST_CASE(define19); // #124
+    TEST_CASE(define20); // #113
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -847,6 +847,16 @@ static void define20() // #113
                   "void foo ( const y < T1 , T2 , T3 , T4 > & x ) { }", preprocess(code));
 }
 
+static void define21() // #66
+{
+    const char code[] = "#define GETMYID(a) ((a))+1\n"
+                        "#define FIGHT_FOO(c, ...) foo(c, ##__VA_ARGS__)\n"
+                        "FIGHT_FOO(1, GETMYID(a));\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "foo ( 1 , ( ( a ) ) + 1 ) ;", preprocess(code));
+}
+
 
 static void define_invalid_1()
 {
@@ -3828,6 +3838,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define18); // #130
     TEST_CASE(define19); // #124
     TEST_CASE(define20); // #113
+    TEST_CASE(define21); // #66
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -783,6 +783,33 @@ static void define16() // #201
                   "\"Blue\" , \"Green\" , \"Purple\" , \"Red\" , \"Yellow\" , \"Orange\" ,", preprocess(code));
 }
 
+static void define17() // #185
+{
+    const char code[] = "#define at(x, y) x##y\n"
+                        "#define b(...) \\\n"
+                        "aa(__VA_ARGS__, , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , \\\n"
+                        ", , , , , , , , 2)\n"
+                        "#define aa(c, d, a, b, e, f, g, h, ab, ac, i, ad, j, k, l, m, n, o, p, ae, q, \\\n"
+                        "r, s, t, u, v, w, x, y, z, af, ag, ah, ai, aj, ak, al, am, an, ao, \\\n"
+                        "ap) \\\n"
+                        "ap\n"
+                        "#define aq(...) ar(b(__VA_ARGS__), __VA_ARGS__) static_assert(true, \" \")\n"
+                        "#define ar(ap, ...) at(I_, ap)(__VA_ARGS__)\n"
+                        "#define I_2(as, a)\n"
+                        "aq(a, array);\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "static_assert ( true , \" \" ) ;", preprocess(code));
+}
 
 
 static void define_invalid_1()
@@ -3761,6 +3788,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define14); // #296
     TEST_CASE(define15); // #231
     TEST_CASE(define16); // #201
+    TEST_CASE(define17); // #185
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -1210,6 +1210,15 @@ static void pragma_backslash()
     ASSERT_EQUALS("", preprocess(code, &outputList));
 }
 
+static void pragma_backslash_2() // #217
+{
+    const char code[] = "#pragma comment(linker, \"foo \\\n"
+                        "bar\")\n";
+
+    simplecpp::OutputList outputList;
+    ASSERT_EQUALS("", preprocess(code, &outputList));
+}
+
 static void dollar()
 {
     ASSERT_EQUALS("$ab", readfile("$ab"));
@@ -3758,6 +3767,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define_va_opt_9); // #632
 
     TEST_CASE(pragma_backslash); // multiline pragma directive
+    TEST_CASE(pragma_backslash_2); // #217
 
     // UB: #ifdef as macro parameter
     TEST_CASE(define_ifdef);

--- a/test.cpp
+++ b/test.cpp
@@ -823,6 +823,18 @@ static void define18() // #130
                   "if ( pGlobalCtx && pGlobalCtx -> debug_level >= 2 ) printf ( \"FT-dbg: \" \"  %02x:%02x:%02x:%02x:%02x:%02x\\n\" , pCtx -> wlan_intf_addr [ i ] [ 0 ] , pCtx -> wlan_intf_addr [ i ] [ 1 ] , pCtx -> wlan_intf_addr [ i ] [ 2 ] , pCtx -> wlan_intf_addr [ i ] [ 3 ] , pCtx -> wlan_intf_addr [ i ] [ 4 ] , pCtx -> wlan_intf_addr [ i ] [ 5 ] ) ;", preprocess(code));
 }
 
+static void define19() // #124
+{
+    const char code[] = "#define CONCAT(tok) tok##suffix\n"
+                        "\n"
+                        "CONCAT(Test);\n"
+                        "CONCAT(const Test);\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "Testsuffix ;\n"
+                  "const Testsuffix ;", preprocess(code));
+}
+
 
 static void define_invalid_1()
 {
@@ -3802,6 +3814,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define16); // #201
     TEST_CASE(define17); // #185
     TEST_CASE(define18); // #130
+    TEST_CASE(define19); // #124
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -719,6 +719,18 @@ static void define13()
                   "}", preprocess(code));
 }
 
+static void define14() // #296
+{
+    const char code[] = "#define bar(x) x % 2\n"
+                        "#define foo(x) printf(#x \"\\n\")\n"
+                        "\n"
+                        " foo(bar(3));\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "printf ( \"bar(3)\" \"\\n\" ) ;", preprocess(code));
+}
+
 
 
 static void define_invalid_1()
@@ -3685,6 +3697,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define11);
     TEST_CASE(define12);
     TEST_CASE(define13);
+    TEST_CASE(define14); // #296
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

--- a/test.cpp
+++ b/test.cpp
@@ -731,6 +731,23 @@ static void define14() // #296
                   "printf ( \"bar(3)\" \"\\n\" ) ;", preprocess(code));
 }
 
+static void define15() // #231
+{
+    const char code[] = "#define CAT(a, b) CAT2(a, b)\n"
+                        "#define CAT2(a, b) a ## b\n"
+                        "#define FOO x\n"
+                        "#define BAR() CAT(F, OO)\n"
+                        "#define BAZ CAT(B, AR)()\n"
+                        "BAZ\n";
+    ASSERT_EQUALS("\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "\n"
+                  "x", preprocess(code));
+}
+
+
 
 
 static void define_invalid_1()
@@ -3698,6 +3715,7 @@ static void runTests(int argc, char **argv, Input input)
     TEST_CASE(define12);
     TEST_CASE(define13);
     TEST_CASE(define14); // #296
+    TEST_CASE(define15); // #231
     TEST_CASE(define_invalid_1);
     TEST_CASE(define_invalid_2);
     TEST_CASE(define_define_1);

